README.md
# BMP

Bitmap generator (bitmap as in BMP image, not bitmap bitset)
  
[![Gem Version](https://badge.fury.io/rb/bmp-ruby.svg)](https://badge.fury.io/rb/bmp-ruby)

## Usage

```
require 'bmp'

# (soon)
```

## Installation

Command Line

```
gem install bmp-ruby
```

Gemfile

```
gem "bmp-ruby", "~>0.1"
```

## Development

(soon)

### BMP File Format

Adapted from: https://en.wikipedia.org/wiki/BMP_file_format

Attribute             |Hex Offset |Dec Offset |Size |Hex Value    |Value                  |Description
----------------------|-----------|-----------|-----|-------------|-----------------------|-----------
BMP Header            |           |           |     |             |                       |
signature             |0h         |0          |2    |42 4D        |"BM"                   |ID field (42h, 4Dh)
file_size             |2h         |2          |4    |9A 00 00 00  |154 bytes (122+32)     |Size of the BMP file
reserved_1            |6h         |6          |2    |00 00        |Unused                 |Application specific
reserved_2            |8h         |8          |2    |00 00        |Unused                 |Application specific
file_offset           |Ah         |10         |4    |7A 00 00 00  |122 bytes (14+108).    |Offset where the pixel array (bitmap data) can be found
DIB Header            |           |           |     |             |                       |
dib_header_size       |Eh         |14         |4    |6C 00 00 00  |108 bytes              |Number of bytes in the DIB header (from this point)
image_width           |12h        |18         |4    |04 00 00 00  |4 pixels               |(left to right order)   Width of the bitmap in pixels
image_height          |16h        |22         |4    |02 00 00 00  |2 pixels               |(bottom to top order)   Height of the bitmap in pixels
planes                |1Ah        |26         |2    |01 00        |1 plane                |Number of color planes being used
bits_per_pixel        |1Ch        |28         |2    |20 00        |32 bits                |Number of bits per pixel
compression           |1Eh        |30         |4    |03 00 00 00  |3                      |BI_BITFIELDS, no pixel array compression used
image_size            |22h        |34         |4    |20 00 00 00  |32 bytes               |Size of the raw bitmap data (including padding)
x_pixels_per_meter    |26h        |38         |4    |13 0B 00 00  |2835 px/meter horiz.   |Print resolution of the image, 72 DPI × 39.3701 inches per meter yields 2834.6472
y_pixels_per_meter    |2Ah        |42         |4    |13 0B 00 00  |2835 px/meter vertical |Print resolution of the image, 72 DPI × 39.3701 inches per meter yields 2834.6472
number_of_colors      |2Eh        |46         |4    |00 00 00 00  |0 colors               |Number of colors in the palette
important_colors      |32h        |50         |4    |00 00 00 00  |0 important colors     |0 means all colors are important
red_channel_bitmask   |36h        |54         |4    |00 00 FF 00  |00FF0000 in big-endian |Red channel bit mask (valid because BI_BITFIELDS is specified)
green_channel_bitmask |3Ah        |58         |4    |00 FF 00 00  |0000FF00 in big-endian |Green channel bit mask (valid because BI_BITFIELDS is specified)
blue_channel_bitmask  |3Eh        |62         |4    |FF 00 00 00  |000000FF in big-endian |Blue channel bit mask (valid because BI_BITFIELDS is specified)
alpha_channel_bitmask |42h        |66         |4    |00 00 00 FF  |FF000000 in big-endian |Alpha channel bit mask
color_space_type      |46h        |70         |4    |20 6E 69 57  |little-endian "Win "   |LCS_WINDOWS_COLOR_SPACE
color_space_endpoints |4Ah        |74         |36   |00 00 00 00\n00 00 00 00\n00 00 00 00\n00 00 00 00\n00 00 00 00\n00 00 00 00\n00 00 00 00\n00 00 00 00|CIEXYZTRIPLE Color Space endpoints |Unused for LCS "Win " or "sRGB"
red_gamma             |6Eh        |110        |4    |00 00 00 00  |0 Red Gamma            |Unused for LCS "Win " or "sRGB"
green_gamma           |72h        |114        |4    |00 00 00 00  |0 Green Gamma          |Unused for LCS "Win " or "sRGB"
blue_gamma            |76h        |118        |4    |00 00 00 00  |0 Blue Gamma           |Unused for LCS "Win " or "sRGB"

### Dev Requirements

* [hoe](https://github.com/seattlerb/hoe) gem manager
* [hoe-bundler] may need `gem install hoe-bundler` installation before using `rake bundler:gemfile`
* [YARD](http://yardoc.org) docs
* [redcarpet](https://github.com/vmg/redcarpet) for yardoc

    [hoe-bundler]: https://github.com/flavorjones/hoe-bundler

### Testing

Tests written with [minitest]

```
rake test
```

  [minitest]: https://github.com/seattlerb/minitest

### Contributing

Send tested code.
Thank you, [contributors]!

  [contributors]: https://github.com/bordeeinc/bmp-ruby/graphs/contributors

### To Do

* (soon)

## License

MIT License

Copyright (c) 2017 Bordee Inc.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

## About

![bordee](http://bordee.com/src/img/surf-with-bordee-github.png)

BMP is maintained and funded by Bordee Inc.
The names and logos for Bordee are trademarks of [Bordee Inc.][bordeeinc]

  [bordeeinc]: http://bordee.com

We love open source software!
See [our other projects][bordee-github]
and [check out Seattle.rb!][community]

  [bordee-github]: https://github.com/bordeeinc
  [community]: https://seattlerb.org


History.txt
=== 0.0.1 / 2017-03-31

* It's alive!

  * Bornday!



test/test_bmp.rb
gem "minitest"
require 'minitest/spec'
require "minitest/autorun"
$:.unshift File.join(File.dirname(__FILE__), '..', 'lib')
require "bmp"
#require 'FileUtils' unless Object.const_defined?('FileUtils')

def clear_dir_except_dot_keep(dir)
  dir_glob = Dir.glob(File.join(dir, '*'))

  FileUtils.rm_rf(dir_glob)
end

def ensure_dir_keep(dir)
  FileUtils.mkdir_p(dir) unless Dir.exist?(dir)
  dir_keep =  File.join(dir, '.keep')
  FileUtils.touch(dir_keep) unless File.exist?(dir_keep)
end

def reset_dir(dir)
  clear_dir_except_dot_keep(dir)
  ensure_dir_keep(dir)
end

class TestBMP < Minitest::Test
  def setup
    @files                = File.join(File.expand_path(File.dirname(__FILE__)), "files")
    @input_dir            = File.join(@files, 'input')
    @output_dir           = File.join(@files, 'output')

    reset_dir(@output_dir)
  end

  def teardown
    reset_dir(@output_dir)
  end

  def test_png_to_bmp_transparent_16
    input_filename  = File.join(@input_dir, 'YVES-framework-transparent-16.png')
    output_filename = File.join(@output_dir, 'YVES-framework-transparent-16.bmp')

    assert File.exist?(input_filename)
    refute File.exist?(output_filename)

    BMP.png_to_bmp(input_filename, output_filename)

    assert File.exist?(output_filename)
  end

  def test_png_to_bmp_transparent_256
    input_filename  = File.join(@input_dir, 'YVES-framework-transparent-256.png')
    output_filename = File.join(@output_dir, 'YVES-framework-transparent-256.bmp')

    assert File.exist?(input_filename)
    refute File.exist?(output_filename)

    BMP.png_to_bmp(input_filename, output_filename)

    assert File.exist?(output_filename)
  end

  def test_png_to_bmp_opaque_16
    input_filename  = File.join(@input_dir, 'YVES-framework-opaque-16.png')
    output_filename = File.join(@output_dir, 'YVES-framework-opaque-16.bmp')

    assert File.exist?(input_filename)
    refute File.exist?(output_filename)

    BMP.png_to_bmp(input_filename, output_filename)

    assert File.exist?(output_filename)
  end

  def test_png_to_bmp_opaque_256
    input_filename  = File.join(@input_dir, 'YVES-framework-opaque-256.png')
    output_filename = File.join(@output_dir, 'YVES-framework-opaque-256.bmp')

    assert File.exist?(input_filename)
    refute File.exist?(output_filename)

    BMP.png_to_bmp(input_filename, output_filename)

    assert File.exist?(output_filename)
  end
end


lib/bmp.rb
require 'core_ext/array/extract_options.rb' unless Array.method_defined?(:extract_options!)
require 'bit-struct'
require 'chunky_png'

require 'bmp/version'
require 'bmp/utils'
require 'bmp/obj'

module BMP
  HEADER_SIZE       = 54
  BITS_PER_PIXEL    = 32

  # make all methods class-methods
  extend self

  # PNG to BMP
  #
  # @param  input_filename  [String] "/path/to/example.png"
  # @param  output_filename [String] "/path/to/example_generated.bmp"
  # @return                 [void]
  #
  def png_to_bmp(input_filename, output_filename)
    bmp = BMP.new(input_filename)

    IO.write(output_filename, bmp)
  end

  #
  # @param  input_filename  [PNG] including path
  # @return                 [BitStruct]
  #
  def new(input_filename)
    img               = ChunkyPNG::Image.from_file(input_filename)
    img_hash          = BMP::Utils.parse_image(img)

    bmp               = BMP::Obj.new
    bmp.file_size     = img_hash[:file_size]
    bmp.image_size    = img_hash[:image_size]
    bmp.image_width   = img_hash[:image_width]
    bmp.image_height  = img_hash[:image_height]
    bmp.pixel_array   = img_hash[:pixel_array]

    bmp
  end
end


lib/bmp/utils.rb
module BMP
  module Utils

    extend self

    # @return [Integer]
    def calc_image_data_size(arg_width, arg_height)
      ((BMP::BITS_PER_PIXEL * arg_width ) / 32.0).ceil * 4 * arg_height
    end

    def int_to_octets_string_endian_little(int)
      [int].pack("V").each_byte.map {|b| b.to_s(16).rjust(2, '0')}.join(':')
    end

    def octets_string_endian_little_to_int(octets_str)
      octets_str.split(':').map {|o| o.to_i(16).chr}.join.unpack('V').first
    end

    # @return [Integer]
    def calc_file_size(img_size, header_size=BMP::HEADER_SIZE)
      img_size + header_size
    end

    def pixel_array_template(arg_width, arg_height)
      Array.new(arg_height) {Array.new(arg_width) {nil}}
    end

    def write_pixel_array(img_px)
      str = ''

      img_px.reverse_each do |row|
        row.each do |color|
          str << pixel_binstring(color)
        end
      end

      return str
    end

    def pixel_binstring(rgba_string)
      raise ArgumentError unless rgba_string =~ /\A\h{8}\z/
      [rgba_string].pack("H*")
    end

    def parse_image(img_in)
      img_width             = img_in.width
      img_height            = img_in.height
      img_data_size         = calc_image_data_size(img_width, img_height)
      img_file_size         = calc_file_size(img_data_size)
      img_pixels            = pixel_array_template(img_width, img_height)

      img_height.times do |row|
        img_width.times do |col|
          rgba = ChunkyPNG::Color.to_hex(img_in.get_pixel(row,col), true)[1..-1]
          rr, gg, bb, aa = rgba.scan(/../)
          img_pixels[col][row] = [bb,gg,rr,aa].join

          #puts row.to_s + ',' + col.to_s + ' => ' + rgba.inspect if @debug
        end
      end

      hash = {}
      hash[:file_size]     = img_file_size
      hash[:image_size]    = img_data_size
      hash[:image_width]   = img_width
      hash[:image_height]  = img_height
      hash[:pixel_array]   = write_pixel_array(img_pixels)

      hash
    end

  end
end


lib/bmp/obj.rb
module BMP
  class Obj < BitStruct
    default_options :endian => :little

    #
    # BMP Header
    #
    char      :signature,             16, 'ID field ["BM" (42h, 4Dh)]',                           :default => 'BM'
    unsigned  :file_size,             32, 'Size of the BMP file [(122+N) bytes]'
    unsigned  :reserved_1,            16, 'Application specific [unused]',                        :default => 0
    unsigned  :reserved_2,            16, 'Application specific [unused]',                        :default => 0

    # Wikipedia:
    #unsigned  :file_offset,           32, 'Offset where the pixel array (bitmap data) can be ' +
    #                                      'found [122 bytes (14+108)]',                           :default => 122
    # IRL:
    unsigned  :file_offset,            32, 'Offset where the pixel array (bitmap data) can be ' +
                                           'found [54 bytes (14+40)]',                             :default => 54

    #
    # DIB Header
    #
    # Wikipedia:
    #unsigned  :dib_header_size,       32,  'Number of bytes in the DIB header (from this point) ' +
    #                                       '[108 bytes]',                                         :default => 108
    # IRL:
    unsigned  :dib_header_size,       32, 'Number of bytes in the DIB header (from this point) ' +
                                          '[40 bytes]',                                            :default => 40

    unsigned  :image_width,           32, 'Width of the bitmap in pixels (left to right order) ' +
                                          '[pixels]'
    unsigned  :image_height,          32, 'Height of the bitmap in pixels (bottom to top order) ' +
                                          '[pixels]'
    unsigned  :planes,                16, 'Number of color planes being used [1 plane]',          :default => 1
    unsigned  :bits_per_pixel,        16, 'Number of bits per pixel [32 bits]',                   :default => 32

    # Wikipedia:
    #unsigned  :compression,           32,  'BI_BITFIELDS, no pixel array compression used [3]',   :default => 3
    # IRL:
    unsigned  :compression,           32, 'BI_BITFIELDS, no pixel array compression used [3]',    :default => 0

    unsigned  :image_size,            32, 'Size of the raw bitmap data (including padding)'
    unsigned  :x_pixels_per_meter,    32, 'Print resolution of the image, 72 DPI × 39.3701 ' +
                                          'inches per meter yields 2834.6472 [2835 pixel/meter ' +
                                          'horizontal]',                                          :default => 2835
    unsigned  :y_pixels_per_meter,    32, 'Print resolution of the image, 72 DPI × 39.3701 ' +
                                          'inches per meter yields 2834.6472 [2835 pixel/meter ' +
                                          'horizontal]',                                          :default => 2835
    unsigned  :number_of_colors,      32, 'Number of colors in the palette [0 colors]',           :default => 0
    unsigned  :important_colors,      32, '0 means all colors are important [0]',                 :default => 0

    # Wikipedia:
    #hex_octets  :red_bitmask,         32,  'Red channel bit mask (valid because BI_BITFIELDS ' +
    #                                       'is specified) [00FF0000 in big-endian]',              :default => "00:00:FF:00", :endian => :big
    #hex_octets  :geen_bitmask,        32,  'Green channel bit mask (valid because BI_BITFIELDS ' +
    #                                       'is specified) [0000FF00 in big-endian]',              :default => "00:FF:00:00", :endian => :big
    #hex_octets  :blue_bitmask,        32,  'Blue channel bit mask (valid because BI_BITFIELDS ' +
    #                                       'is specified) [000000FF in big-endian]',              :default => "FF:00:00:00", :endian => :big
    #hex_octets  :alpha_bitmask,       32,  'Alpha channel bit mask [FF000000 in big-endian]',     :default => "00:00:00:FF", :endian => :big
    #char        :color_space,         32,  'LCS_WINDOWS_COLOR_SPACE [little-endian "Win "]',      :default => 'Win '
    #unsigned    :cs_endpoints,        288, 'CIEXYZTRIPLE Color Space endpoints [unused for LCS ' +
    #                                       '"Win " or "sRGB"]',                                   :default => 0
    #unsigned    :red_gamma,           32,  'Red Gamma - Unused for LCS "Win " or "sRGB"',         :default => 0
    #unsigned    :green_gamma,         32,  'Green Gamma - Unused for LCS "Win " or "sRGB"',       :default => 0
    #unsigned    :blue_gamma,          32,  'Blue Gamma - Unused for LCS "Win " or "sRGB"',        :default => 0
    # IRL:
    #nil

    rest      :pixel_array,               'Image Data' 
    note                                  '     rest is application defined image data'
  end
end
